### ISSUE TITLE
Bug: Error on attribute deletion after updating to v2.4.161

### ISSUE THREAD

**Issue by samitainio** [MISP/MISP#8555]
### Expected behavior

I can delete an attribute via the user interface.

### Actual behavior

The log file at `/var/www/MISP/app/tmp/logs/error.log` gets an error message and the attribute is not deleted from the event.

### Steps to reproduce

1. Create test event.
2. Add an attribute. I used domain.com
3. Try to delete the attribute via the UI.

### Version

Currently installed version… v2.4.161 (4d9ea1e4e37eddb4d2974d27e3f64c5f7853fa43)

### Operating System

Ubuntu

### Operating System version

20.04

### PHP version

7.4.3

### Relevant log output

```shell
2022-08-20 23:55:50 Error: [PDOException] SQLSTATE[42S22]: Column not found: 1054 Unknown column 'Correlation.object_id' in 'field list'
Request URL: /attributes/delete/18320070/true
Stack Trace:
#0 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Datasource/DboSource.php(502): PDOStatement->execute()
#1 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Datasource/DboSource.php(468): DboSource->_execute()
#2 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Datasource/DboSource.php(715): DboSource->execute()
#3 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Datasource/DboSource.php(1564): DboSource->fetchAll()
#4 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Datasource/DboSource.php(1398): DboSource->_fetchHasMany()
#5 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Datasource/DboSource.php(1257): DboSource->queryAssociation()
#6 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Model.php(3053): DboSource->read()
#7 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Model.php(3025): Model->_readDataSource()
#8 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Model.php(1636): Model->find()
#9 /var/www/MISP/app/Model/Attribute.php(512): Model->read()
#10 /var/www/MISP/app/Lib/Tools/BetterCakeEventManager.php(21): Attribute->beforeDelete()
#11 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Model.php(2694): BetterCakeEventManager->dispatch()
#12 /var/www/MISP/app/Model/Attribute.php(3063): Model->delete()
#13 /var/www/MISP/app/Controller/AttributesController.php(1060): Attribute->deleteAttribute()
#14 [internal function]: AttributesController->delete()
#15 /var/www/MISP/app/Lib/cakephp/lib/Cake/Controller/Controller.php(499): ReflectionMethod->invokeArgs()
#16 /var/www/MISP/app/Lib/cakephp/lib/Cake/Routing/Dispatcher.php(193): Controller->invokeAction()
#17 /var/www/MISP/app/Lib/cakephp/lib/Cake/Routing/Dispatcher.php(167): Dispatcher->_invoke()
#18 /var/www/MISP/app/webroot/index.php(99): Dispatcher->dispatch()
#19 {main}
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by github-germ** [comment]
Not sure what your PyMISP is setting; however, I do see that the {{correlations}} table does not contain an {{object_id}} column with 2.4.161:
```
mysql> desc correlations;
+--------------------+---------+------+-----+---------+----------------+
| Field              | Type    | Null | Key | Default | Extra          |
+--------------------+---------+------+-----+---------+----------------+
| id                 | int     | NO   | PRI | NULL    | auto_increment |
| value              | text    | NO   |     | NULL    |                |
| 1_event_id         | int     | NO   | MUL | NULL    |                |
| 1_attribute_id     | int     | NO   | MUL | NULL    |                |
| event_id           | int     | NO   | MUL | NULL    |                |
| attribute_id       | int     | NO   | MUL | NULL    |                |
| org_id             | int     | NO   |     | NULL    |                |
| distribution       | tinyint | NO   |     | NULL    |                |
| a_distribution     | tinyint | NO   |     | NULL    |                |
| sharing_group_id   | int     | NO   |     | NULL    |                |
| a_sharing_group_id | int     | NO   |     | NULL    |                |
| date               | date    | NO   |     | NULL    |                |
| info               | text    | NO   |     | NULL    |                |
+--------------------+---------+------+-----+---------+----------------+
13 rows in set (0.03 sec)

mysql>
```

**Comment by iglocska** [comment]
Will have a look tomorrow - the correlations table is deprecated, so that's not the issue (`Correlation.` is an alias for the table of the correlation model, which will in most cases be the default_correlations table).

**Comment by androssfox** [comment]
Having the same issue, it fails to delete attributes from the Web UI (same 'Correlation.object_id' in 'field list' error as **samitainio**), however it works using API.

We are using v2.4.161 (https://github.com/MISP/MISP/commit/5d4f73fcbbef577e39390cba0e38d7281de6cc4e)

**Comment by HugeekMcGill** [comment]
Hi,

I have the same issue with 2.4.161

Looks like it's a migration issue. I passed from 159 to 161 and I have this error. 

I deploy with a new setup and i have no issues.

Any updated ? I can provide information if needed.

**Comment by samitainio** [comment]
@iglocska let me know if you need additional information? Are there any updates or ideas for how we could circumvent this temporarily while we wait for a proper fix?

**Comment by cudeso** [comment]
@samitainio  had the same problem on 2.4.161 
```2022-09-12 12:37:20 Error: [PDOException] SQLSTATE[42S22]: Column not found: 1054 Unknown column 'Correlation.object_id' in 'field list'```

Upgrading to **2.4.162** solved it for me.

**Comment by cudeso** [comment]
> @samitainio had the same problem on 2.4.161 `2022-09-12 12:37:20 Error: [PDOException] SQLSTATE[42S22]: Column not found: 1054 Unknown column 'Correlation.object_id' in 'field list'`
> 
> Upgrading to **2.4.162** solved it for me.

Actually the upgrade did **not** solve it. It worked on one attribute. But for one or the other reason, it occurred when deleting other attributes.

**Comment by cudeso** [comment]
Attempt 2.
- Running 'clean model cache' allows me to do "delete attributes" once,
- Re-opening the event and deleting attributes gives the above error
- Re-running 'clean model cache' then allows me one time to delete attributes

**Comment by mokaddem** [comment]
Interesting behavior observed by @cudeso.. Unfortunately, I could not reproduce it even on the provided commit id.
And cleaning the cache should have persistently solved the problem.
Is anyone experiencing this as well? Or does cleaning the cache solve (`/servers/serverSettings/diagnostics` -> `clear cache`) the issue (as it should).

**Comment by cudeso** [comment]
Hi @mokaddem ; I dived a bit deeper and have additional info. The environment where I did the tests is a setup that has a lot of "merged" events. 
When deleting attributes from newly created events, there is no problem. Deleting works as expected. The error only occurs when deleting attributes in events which are merged. I can solve this for a "one time delete" by cleaning the cache, however after one delete the problem re-occurs.

The setup isn't critical; happy to do some additional tests or provide you debug output when needed.

**Comment by samitainio** [comment]
@iglocska @mokaddem @cudeso I am seeing this error also with the newly created events. It seems, like already discussed but just to confirm from my instance, that the clean model cache works for one time deletion and if I attempt to delete another attribute then the error comes back and I need to rerun the clean model cache to be able to delete another attribute. However my instance does also have merged events but I haven't seen any difference in behaviour between the merged events and the new events.

I'm currently running MISP with the updated version v2.4.162 (6460092dede7b3c58ffa38aaaca52b853afd6008) and it is still showing the same symptoms as originally reported.

**Comment by mokaddem** [comment]
Most probably unrelated but I just fixed another issue preventing attribute hard deletion while being a user with the role `org_admin`.
Still can't reproduce the original problem though.

**Comment by cudeso** [comment]
@mokaddem what's the commit ID?

**Comment by cudeso** [comment]
@mokaddem For the events where I did not had the problem, the soft-delete works OK; however doing the hard-delete gives the error. The user is an `admin` (site admin).

**Comment by mokaddem** [comment]
I tried the commit provided by @samitainio https://github.com/MISP/MISP/commit/6460092dede7b3c58ffa38aaaca52b853afd6008
I didn't have the issues while using the `admin` role, that's why I tried to use the `org_admin` and noticed the other issue.

**Comment by cudeso** [comment]
The "merged" events had nothing to do with it. The problem lies with the hard-delete.

When printing out the SQL commands from `app/Lib/cakephp/lib/Cake/Model/Datasource/DboSource.php` (function execute):

1/ After clean cache. A hard delete without errors gives a bunch of errors, among with this one using the object_id. It sets Correlation to **default_correlations**.
```
SELECT `Correlation`.`id`, `Correlation`.`attribute_id`, `Correlation`.`object_id`, `Correlation`.`event_id`, `Correlation`.`org_id`, `Correlation`.`distribution`, `Correlation`.`object_distribution`, `Correlation`.`event_distribution`, `Correlation`.`sharing_group_id`, `Correlation`.`object_sharing_group_id`, `Correlation`.`event_sharing_group_id`, `Correlation`.`1_attribute_id`, `Correlation`.`1_object_id`, `Correlation`.`1_event_id`, `Correlation`.`1_org_id`, `Correlation`.`1_distribution`, `Correlation`.`1_object_distribution`, `Correlation`.`1_event_distribution`, `Correlation`.`1_sharing_group_id`, `Correlation`.`1_object_sharing_group_id`, `Correlation`.`1_event_sharing_group_id`, `Correlation`.`value_id` FROM `misp`.`default_correlations` AS `Correlation` WHERE `Correlation`.`attribute_id` = (2295094)
```

The SQL commands after the previous command is 
```
SELECT `Sighting`.`id`, `Sighting`.`attribute_id`, `Sighting`.`event_id`, `Sighting`.`org_id`, `Sighting`.`date_sighting`, `Sighting`.`uuid`, `Sighting`.`source`, `Sighting`.`type` FROM `misp`.`sightings` AS `Sighting` WHERE `Sighting`.`attribute_id` = (2295094)
DELETE `Correlation` FROM `misp`.`default_correlations` AS `Correlation` WHERE ((`1_attribute_id` = 2295094) OR (`attribute_id` = 2295094))
```

2/ Trying a second delete, gives the error *Unknown column 'Correlation.object_id' in 'field list'*. It sets Correlation to **correlations**. This table does not have the object_id column.

```
SELECT `Correlation`.`id`, `Correlation`.`attribute_id`, `Correlation`.`object_id`, `Correlation`.`event_id`, `Correlation`.`org_id`, `Correlation`.`distribution`, `Correlation`.`object_distribution`, `Correlation`.`event_distribution`, `Correlation`.`sharing_group_id`, `Correlation`.`object_sharing_group_id`, `Correlation`.`event_sharing_group_id`, `Correlation`.`1_attribute_id`, `Correlation`.`1_object_id`, `Correlation`.`1_event_id`, `Correlation`.`1_org_id`, `Correlation`.`1_distribution`, `Correlation`.`1_object_distribution`, `Correlation`.`1_event_distribution`, `Correlation`.`1_sharing_group_id`, `Correlation`.`1_object_sharing_group_id`, `Correlation`.`1_event_sharing_group_id`, `Correlation`.`value_id` FROM `misp`.`correlations` AS `Correlation` WHERE `Correlation`.`attribute_id` = (2295090)
```

**Comment by samitainio** [comment]
@iglocska @cudeso @mokaddem Just a quick update from this end. I've now updated to the latest release version v2.4.163 (16bcf9fcdcf6d2c800060a81c0b84fbaa5824feb). Issue is still relevant and the symptoms are the same.

**Comment by skalvaro** [comment]
Seem to encounter the same on  v2.4.161 (47674a80511f058ea35fff0a423b62d69e7139f9) when trying to delete attributes or accept/delete proposals.
Is there any work-around known or should we upgrade to v2.4.163 (16bcf9fcdcf6d2c800060a81c0b84fbaa5824feb)?

**Comment by cudeso** [comment]
Initial tests with the upgrade to v2.4.164 indicate the problem seemed to be solved.

@samitainio are you able to upgrade and verify?

**Comment by samitainio** [comment]
@cudeso upgraded & verified: it seems to work now. I don't know what was changed but I don't see errors at this time.

**Comment by samitainio** [comment]
@androssfox @HugeekMcGill @skalvaro could you also comment if this issue is resolved for your instances with the latest MISP release v2.4.164? If it is working after the update I will then proceed to close this issue.

**Comment by eowyn263** [comment]
@samitainio @cudeso The upgrade to v2.4.164 solved the issue for hard delete, but I'm still getting this error when accepting the proposal (clean cache works only on one attribute too in this case). Are you also still getting this error when accepting the proposal?

**Comment by HugeekMcGill** [comment]
@samitainio yes the issue is solved on my side.

Since I'm using containers and importing misp_db, I had to manually add the missing MySQL colums in misp_db to trigger the rest of the update script.

If that can save someone else time, ping me if you have the same issue, it took my alot of time to figure out the solution.

**Comment by dragsu** [comment]
As @skalvaro noticed, I had the same issue with the proposals and it is resolved in v2.4.168.
